File: backend/services/analysis_engine.go
package services

import (
	"fmt"
	"math"
	"sort"
	"strconv"
	"strings"

	"github.com/yourname/go-fiber-backend/backend/models"
)

// AnalysisEngine processes analysis requests and generates chart data
type AnalysisEngine struct {
	gemini *GeminiService
}

// NewAnalysisEngine creates a new analysis engine
func NewAnalysisEngine(gemini *GeminiService) *AnalysisEngine {
	return &AnalysisEngine{
		gemini: gemini,
	}
}

// ProcessAnalysisResult contains the result of analysis processing
type ProcessAnalysisResult struct {
	Insights      string
	Charts        []models.ChartConfig
	ChartStatus   string // "success", "partial", "failed", "not_feasible"
	ChartMessage  string
	RetryAttempts int
}

// ProcessAnalysis processes analysis request using two-step flow with retry logic
func (ae *AnalysisEngine) ProcessAnalysis(dataset *models.Dataset, prompt string) (*ProcessAnalysisResult, error) {
	// Validate inputs
	if dataset == nil {
		return nil, fmt.Errorf("dataset cannot be nil")
	}
	if len(dataset.Rows) == 0 {
		return nil, fmt.Errorf("dataset is empty")
	}
	if prompt == "" {
		return nil, fmt.Errorf("prompt cannot be empty")
	}

	// Step 1: Pure reasoning (no charts)
	insights, err := ae.gemini.Analyze(dataset, prompt)
	if err != nil {
		return nil, fmt.Errorf("analysis step failed: %w", err)
	}
	
	// Validate insights were generated
	if insights == "" {
		return nil, fmt.Errorf("analysis step returned empty insights")
	}

	// Step 2: Chart generation with retry logic
	result := &ProcessAnalysisResult{
		Insights:      insights,
		Charts:        []models.ChartConfig{},
		ChartStatus:   "failed",
		ChartMessage:  "",
		RetryAttempts: 0,
	}

	// Check feasibility before attempting chart generation
	if !ae.isChartGenerationFeasible(dataset) {
		result.ChartStatus = "not_feasible"
		result.ChartMessage = "Graph cannot be generated due to insufficient or incompatible data. The dataset may be too small, missing required columns, or lack numeric data."
		return result, nil
	}

	// Retry loop (max 3 attempts)
	maxAttempts := 3
	var lastError string

	for attempt := 1; attempt <= maxAttempts; attempt++ {
		result.RetryAttempts = attempt

		chartConfigs, err := ae.gemini.GenerateCharts(dataset, prompt, lastError)
		if err != nil {
			// Check if it's a JSON parsing error
			lastError = err.Error()

			// If it's a JSON error and we have retries left, continue
			if attempt < maxAttempts {
				// Enhance error message for next retry
				if strings.Contains(err.Error(), "JSON") || strings.Contains(err.Error(), "json") {
					lastError = fmt.Sprintf("Previous response did NOT include valid graph JSON. Error: %s. You MUST return ONLY valid JSON in the exact format specified, with no explanations, markdown, or code blocks.", err.Error())
				}
				continue // Retry
			}
			// Final attempt failed
			result.ChartStatus = "failed"
			result.ChartMessage = "Visual representation could not be generated reliably for this prompt or your requirements."
			return result, nil
		}

		// Validate and populate chart data
		validCharts := []models.ChartConfig{}
		for i := range chartConfigs {
			chart := &chartConfigs[i]

			// Validate chart configuration
			if !ae.validateChartConfig(dataset, chart) {
				continue // Skip invalid chart
			}

			// Generate chart data
			chart.Data = ae.generateChartData(dataset, chart)

			// Check if data was generated successfully
			if len(chart.Data) == 0 {
				continue // Skip charts with no data
			}

			validCharts = append(validCharts, *chart)
		}

		if len(validCharts) > 0 {
			result.Charts = validCharts
			if len(validCharts) == len(chartConfigs) {
				result.ChartStatus = "success"
			} else {
				result.ChartStatus = "partial"
				result.ChartMessage = fmt.Sprintf("Generated %d out of %d requested charts", len(validCharts), len(chartConfigs))
			}
			return result, nil
		}

		// No valid charts generated, prepare retry
		if attempt < maxAttempts {
			lastError = "No valid charts were generated. Please ensure column names match the schema exactly and chart types are appropriate for the data."
			continue
		}
	}

	// All attempts failed
	result.ChartStatus = "failed"
	result.ChartMessage = "Visual representation could not be generated reliably for this prompt or your requirements."
	return result, nil
}

// isChartGenerationFeasible checks if chart generation is feasible
func (ae *AnalysisEngine) isChartGenerationFeasible(dataset *models.Dataset) bool {
	// Check dataset size
	if len(dataset.Rows) <= 1 {
		return false
	}

	// At least one column required
	if len(dataset.Columns) == 0 {
		return false
	}

	// For basic charts like pie charts, we can work with categorical data only
	// But for most chart types, we need at least one numeric column
	// We'll allow it if there's at least one column (categorical charts can work)
	return true
}

// validateChartConfig validates a chart configuration against the dataset
func (ae *AnalysisEngine) validateChartConfig(dataset *models.Dataset, chart *models.ChartConfig) bool {
	// Check required columns exist
	columnMap := make(map[string]bool)
	for _, col := range dataset.Columns {
		columnMap[col.Name] = true
	}

	switch chart.Type {
	case "bar", "line", "area", "scatter", "combo":
		if chart.X != "" && !columnMap[chart.X] {
			return false
		}
		if chart.Y != "" && !columnMap[chart.Y] {
			return false
		}
		if chart.Y2 != "" && !columnMap[chart.Y2] {
			return false
		}
		if chart.GroupBy != "" && !columnMap[chart.GroupBy] {
			return false
		}
	case "pie":
		if chart.Category != "" && !columnMap[chart.Category] {
			return false
		}
		if chart.Value != "" && !columnMap[chart.Value] {
			return false
		}
	case "histogram":
		if chart.X != "" && !columnMap[chart.X] {
			return false
		}
	case "boxplot":
		if chart.Y != "" && !columnMap[chart.Y] {
			return false
		}
	case "bubble":
		if chart.X != "" && !columnMap[chart.X] {
			return false
		}
		if chart.Y != "" && !columnMap[chart.Y] {
			return false
		}
		if chart.Z != "" && !columnMap[chart.Z] {
			return false
		}
	case "correlation", "heatmap":
		if len(chart.Columns) == 0 {
			return false
		}
		for _, col := range chart.Columns {
			if !columnMap[col] {
				return false
			}
		}
	default:
		return false // Unknown chart type
	}

	return true
}

// generateChartData generates data points for a chart
func (ae *AnalysisEngine) generateChartData(dataset *models.Dataset, chart *models.ChartConfig) []map[string]interface{} {
	data := make([]map[string]interface{}, 0)

	switch chart.Type {
	case "bar", "line", "area":
		if chart.X != "" && chart.Y != "" {
			if chart.GroupBy != "" {
				// Grouped or stacked bar chart
				data = ae.generateGroupedData(dataset, chart)
			} else if chart.Aggregate != "" {
				// Aggregated data
				data = ae.generateAggregatedData(dataset, chart)
			} else {
				// Simple XY data
				data = ae.generateXYData(dataset, chart.X, chart.Y)
			}
		}
	case "scatter":
		if chart.X != "" && chart.Y != "" {
			data = ae.generateXYData(dataset, chart.X, chart.Y)
		}
	case "bubble":
		if chart.X != "" && chart.Y != "" && chart.Z != "" {
			data = ae.generateBubbleData(dataset, chart)
		}
	case "pie":
		if chart.Category != "" && chart.Value != "" {
			data = ae.generatePieData(dataset, chart.Category, chart.Value)
		}
	case "combo":
		if chart.X != "" && chart.Y != "" {
			data = ae.generateComboData(dataset, chart)
		}
	case "histogram":
		if chart.X != "" {
			data = ae.generateHistogramData(dataset, chart)
		}
	case "boxplot":
		if chart.Y != "" {
			data = ae.generateBoxplotData(dataset, chart)
		}
	case "correlation", "heatmap":
		if len(chart.Columns) > 0 {
			data = ae.generateCorrelationData(dataset, chart)
		}
	}

	return data
}

func (ae *AnalysisEngine) generateXYData(dataset *models.Dataset, xCol, yCol string) []map[string]interface{} {
	xValues := dataset.GetColumnData(xCol)
	yValues := dataset.GetColumnData(yCol)

	if len(xValues) != len(yValues) {
		return []map[string]interface{}{}
	}

	data := make([]map[string]interface{}, 0, len(xValues))
	for i := 0; i < len(xValues) && i < 100; i++ { // Limit to 100 points
		xVal := xValues[i]
		yVal := yValues[i]

		// Try to convert to numbers
		xNum, xErr := ae.toNumber(xVal)
		yNum, yErr := ae.toNumber(yVal)

		point := map[string]interface{}{
			xCol: xVal,
			yCol: yVal,
		}

		if xErr == nil {
			point[xCol] = xNum
		}
		if yErr == nil {
			point[yCol] = yNum
		}

		data = append(data, point)
	}

	return data
}

func (ae *AnalysisEngine) generatePieData(dataset *models.Dataset, categoryCol, valueCol string) []map[string]interface{} {
	categoryValues := dataset.GetColumnData(categoryCol)
	valueValues := dataset.GetColumnData(valueCol)

	if len(categoryValues) != len(valueValues) {
		return []map[string]interface{}{}
	}

	// Aggregate by category
	aggregated := make(map[string]float64)
	for i := 0; i < len(categoryValues); i++ {
		category := fmt.Sprintf("%v", categoryValues[i])
		value, err := ae.toNumber(valueValues[i])
		if err == nil {
			aggregated[category] += value
		}
	}

	data := make([]map[string]interface{}, 0, len(aggregated))
	for category, value := range aggregated {
		data = append(data, map[string]interface{}{
			categoryCol: category,
			valueCol:    value,
		})
	}

	return data
}

// generateGroupedData generates data for grouped/stacked bar charts
func (ae *AnalysisEngine) generateGroupedData(dataset *models.Dataset, chart *models.ChartConfig) []map[string]interface{} {
	xValues := dataset.GetColumnData(chart.X)
	yValues := dataset.GetColumnData(chart.Y)
	groupValues := dataset.GetColumnData(chart.GroupBy)

	if len(xValues) != len(yValues) || len(xValues) != len(groupValues) {
		return []map[string]interface{}{}
	}

	// Group data by X and GroupBy
	grouped := make(map[string]map[string][]float64) // xValue -> groupValue -> []values (for avg calculation)
	groups := make(map[string]bool)

	// First pass: collect all values
	for i := 0; i < len(xValues); i++ {
		xVal := fmt.Sprintf("%v", xValues[i])
		groupVal := fmt.Sprintf("%v", groupValues[i])
		yVal, err := ae.toNumber(yValues[i])
		if err != nil {
			continue
		}

		if grouped[xVal] == nil {
			grouped[xVal] = make(map[string][]float64)
		}

		if grouped[xVal][groupVal] == nil {
			grouped[xVal][groupVal] = make([]float64, 0)
		}

		grouped[xVal][groupVal] = append(grouped[xVal][groupVal], yVal)
		groups[groupVal] = true
	}

	// Second pass: apply aggregation
	data := make([]map[string]interface{}, 0)
	groupList := make([]string, 0, len(groups))
	for g := range groups {
		groupList = append(groupList, g)
	}

	for xVal, groupMap := range grouped {
		point := map[string]interface{}{
			chart.X: xVal,
		}

		for _, groupVal := range groupList {
			values := groupMap[groupVal]
			if len(values) == 0 {
				point[groupVal] = 0
				continue
			}

			var result float64
			switch chart.Aggregate {
			case "sum", "":
				result = 0
				for _, v := range values {
					result += v
				}
			case "avg":
				result = 0
				for _, v := range values {
					result += v
				}
				result = result / float64(len(values))
			case "count":
				result = float64(len(values))
			case "max":
				result = values[0]
				for _, v := range values {
					if v > result {
						result = v
					}
				}
			case "min":
				result = values[0]
				for _, v := range values {
					if v < result {
						result = v
					}
				}
			default:
				result = 0
				for _, v := range values {
					result += v
				}
			}

			point[groupVal] = result
		}

		data = append(data, point)
	}

	return data
}

// generateAggregatedData generates aggregated data (sum, avg, count, etc.)
func (ae *AnalysisEngine) generateAggregatedData(dataset *models.Dataset, chart *models.ChartConfig) []map[string]interface{} {
	xValues := dataset.GetColumnData(chart.X)
	yValues := dataset.GetColumnData(chart.Y)

	if len(xValues) != len(yValues) {
		return []map[string]interface{}{}
	}

	// Aggregate by X
	aggregated := make(map[string][]float64)

	for i := 0; i < len(xValues); i++ {
		xVal := fmt.Sprintf("%v", xValues[i])
		yVal, err := ae.toNumber(yValues[i])
		if err != nil {
			continue
		}

		if aggregated[xVal] == nil {
			aggregated[xVal] = make([]float64, 0)
		}
		aggregated[xVal] = append(aggregated[xVal], yVal)
	}

	// Apply aggregation function
	data := make([]map[string]interface{}, 0, len(aggregated))
	for xVal, values := range aggregated {
		var result float64
		switch chart.Aggregate {
		case "sum":
			result = 0
			for _, v := range values {
				result += v
			}
		case "avg":
			result = 0
			for _, v := range values {
				result += v
			}
			result = result / float64(len(values))
		case "count":
			result = float64(len(values))
		case "max":
			result = values[0]
			for _, v := range values {
				if v > result {
					result = v
				}
			}
		case "min":
			result = values[0]
			for _, v := range values {
				if v < result {
					result = v
				}
			}
		default:
			result = 0
			for _, v := range values {
				result += v
			}
		}

		data = append(data, map[string]interface{}{
			chart.X: xVal,
			chart.Y: result,
		})
	}

	return data
}

// generateComboData generates data for combo charts (bar + line)
func (ae *AnalysisEngine) generateComboData(dataset *models.Dataset, chart *models.ChartConfig) []map[string]interface{} {
	xValues := dataset.GetColumnData(chart.X)
	yValues := dataset.GetColumnData(chart.Y)
	y2Values := dataset.GetColumnData(chart.Y2)

	if len(xValues) != len(yValues) {
		return []map[string]interface{}{}
	}

	data := make([]map[string]interface{}, 0, len(xValues))
	for i := 0; i < len(xValues) && i < 100; i++ {
		xVal := xValues[i]
		yVal, yErr := ae.toNumber(yValues[i])

		point := map[string]interface{}{
			chart.X: xVal,
			chart.Y: yVal,
		}

		if chart.Y2 != "" && i < len(y2Values) {
			y2Val, y2Err := ae.toNumber(y2Values[i])
			if y2Err == nil {
				point[chart.Y2] = y2Val
			}
		}

		if yErr == nil {
			data = append(data, point)
		}
	}

	return data
}

// generateBubbleData generates data for bubble charts
func (ae *AnalysisEngine) generateBubbleData(dataset *models.Dataset, chart *models.ChartConfig) []map[string]interface{} {
	xValues := dataset.GetColumnData(chart.X)
	yValues := dataset.GetColumnData(chart.Y)
	zValues := dataset.GetColumnData(chart.Z)

	if len(xValues) != len(yValues) || len(xValues) != len(zValues) {
		return []map[string]interface{}{}
	}

	data := make([]map[string]interface{}, 0, len(xValues))
	for i := 0; i < len(xValues) && i < 100; i++ {
		xVal, xErr := ae.toNumber(xValues[i])
		yVal, yErr := ae.toNumber(yValues[i])
		zVal, zErr := ae.toNumber(zValues[i])

		if xErr == nil && yErr == nil && zErr == nil {
			data = append(data, map[string]interface{}{
				chart.X: xVal,
				chart.Y: yVal,
				chart.Z: zVal,
			})
		}
	}

	return data
}

// generateHistogramData generates histogram data
func (ae *AnalysisEngine) generateHistogramData(dataset *models.Dataset, chart *models.ChartConfig) []map[string]interface{} {
	values := dataset.GetColumnData(chart.X)

	// Extract numeric values
	numbers := make([]float64, 0)
	for _, val := range values {
		if num, err := ae.toNumber(val); err == nil {
			numbers = append(numbers, num)
		}
	}

	if len(numbers) == 0 {
		return []map[string]interface{}{}
	}

	// Determine bins
	bins := chart.Bins
	if bins == 0 {
		bins = 20 // Default
	}

	// Find min and max
	min, max := numbers[0], numbers[0]
	for _, n := range numbers {
		if n < min {
			min = n
		}
		if n > max {
			max = n
		}
	}

	// Create bins
	binWidth := (max - min) / float64(bins)
	binCounts := make([]int, bins)

	// Count values in each bin
	for _, n := range numbers {
		binIndex := int((n - min) / binWidth)
		if binIndex >= bins {
			binIndex = bins - 1
		}
		if binIndex < 0 {
			binIndex = 0
		}
		binCounts[binIndex]++
	}

	// Create data points
	data := make([]map[string]interface{}, 0, bins)
	for i := 0; i < bins; i++ {
		binStart := min + float64(i)*binWidth
		binEnd := min + float64(i+1)*binWidth
		data = append(data, map[string]interface{}{
			"bin":   fmt.Sprintf("%.2f-%.2f", binStart, binEnd),
			"count": binCounts[i],
			"start": binStart,
			"end":   binEnd,
		})
	}

	return data
}

// generateBoxplotData generates boxplot data
func (ae *AnalysisEngine) generateBoxplotData(dataset *models.Dataset, chart *models.ChartConfig) []map[string]interface{} {
	values := dataset.GetColumnData(chart.Y)

	// Extract numeric values
	numbers := make([]float64, 0)
	for _, val := range values {
		if num, err := ae.toNumber(val); err == nil {
			numbers = append(numbers, num)
		}
	}

	if len(numbers) == 0 {
		return []map[string]interface{}{}
	}

	// Sort numbers
	sort.Float64s(numbers)

	// Calculate quartiles
	n := len(numbers)
	q1Idx := n / 4
	medianIdx := n / 2
	q3Idx := 3 * n / 4

	q1 := numbers[q1Idx]
	median := numbers[medianIdx]
	q3 := numbers[q3Idx]

	// Calculate IQR and outliers
	iqr := q3 - q1
	lowerFence := q1 - 1.5*iqr
	upperFence := q3 + 1.5*iqr

	// Find outliers
	outliers := make([]float64, 0)
	for _, n := range numbers {
		if n < lowerFence || n > upperFence {
			outliers = append(outliers, n)
		}
	}

	// Min and max within fences
	minVal := numbers[0]
	maxVal := numbers[n-1]
	for _, n := range numbers {
		if n >= lowerFence && n < minVal {
			minVal = n
		}
		if n <= upperFence && n > maxVal {
			maxVal = n
		}
	}

	return []map[string]interface{}{
		{
			"min":      minVal,
			"q1":       q1,
			"median":   median,
			"q3":       q3,
			"max":      maxVal,
			"outliers": outliers,
		},
	}
}

// generateCorrelationData generates correlation matrix data
func (ae *AnalysisEngine) generateCorrelationData(dataset *models.Dataset, chart *models.ChartConfig) []map[string]interface{} {
	if len(chart.Columns) == 0 {
		return []map[string]interface{}{}
	}

	// Get data for all columns
	columnData := make(map[string][]float64)
	for _, colName := range chart.Columns {
		values := dataset.GetColumnData(colName)
		numbers := make([]float64, 0)
		for _, val := range values {
			if num, err := ae.toNumber(val); err == nil {
				numbers = append(numbers, num)
			}
		}
		columnData[colName] = numbers
	}

	// Calculate correlation matrix
	data := make([]map[string]interface{}, 0)
	for _, col1 := range chart.Columns {
		row := map[string]interface{}{
			"column": col1,
		}
		for _, col2 := range chart.Columns {
			corr := ae.calculateCorrelation(columnData[col1], columnData[col2])
			row[col2] = corr
		}
		data = append(data, row)
	}

	return data
}

// calculateCorrelation calculates Pearson correlation coefficient
func (ae *AnalysisEngine) calculateCorrelation(x, y []float64) float64 {
	if len(x) != len(y) || len(x) == 0 {
		return 0
	}

	// Find common length
	minLen := len(x)
	if len(y) < minLen {
		minLen = len(y)
	}
	if minLen == 0 {
		return 0
	}

	// Calculate means
	var sumX, sumY float64
	for i := 0; i < minLen; i++ {
		sumX += x[i]
		sumY += y[i]
	}
	meanX := sumX / float64(minLen)
	meanY := sumY / float64(minLen)

	// Calculate correlation
	var numerator, sumSqX, sumSqY float64
	for i := 0; i < minLen; i++ {
		dx := x[i] - meanX
		dy := y[i] - meanY
		numerator += dx * dy
		sumSqX += dx * dx
		sumSqY += dy * dy
	}

	denominator := math.Sqrt(sumSqX * sumSqY)
	if denominator == 0 {
		return 0
	}

	return numerator / denominator
}

func (ae *AnalysisEngine) toNumber(val interface{}) (float64, error) {
	switch v := val.(type) {
	case float64:
		return v, nil
	case int:
		return float64(v), nil
	case string:
		return strconv.ParseFloat(v, 64)
	default:
		return strconv.ParseFloat(fmt.Sprintf("%v", v), 64)
	}
}

File: backend-python/services/analysis_engine.py
from __future__ import annotations
import json
import logging
import math
from collections import defaultdict
from typing import Any

import numpy as np

from models.api import AnalyzeResponse, ChartConfig, ColumnInfo
from models.dataset import Dataset
from services import gemini

logger = logging.getLogger(__name__)

_MAX_RETRIES = 3
_MAX_ROWS_FOR_CHART = 500


def _build_dataset_context(dataset: Dataset) -> str:
    col_desc = ", ".join(f"{c.name} ({c.type})" for c in dataset.columns)
    summary_lines = []
    for col_name, stats in dataset.summary.items():
        col = next((c for c in dataset.columns if c.name == col_name), None)
        if col and col.type == "number":
            summary_lines.append(
                f"  {col_name}: min={stats.min}, max={stats.max}, mean={stats.mean:.2f}, stdDev={stats.std_dev:.2f}"
            )
        else:
            summary_lines.append(f"  {col_name}: {stats.unique_count} unique values, {stats.null_count} nulls")
    summary_text = "\n".join(summary_lines)
    return f"Columns: {col_desc}\nTotal rows: {len(dataset.rows)}\nStatistics:\n{summary_text}"


def _valid_column(dataset: Dataset, name: str | None) -> bool:
    if not name:
        return False
    return any(c.name == name for c in dataset.columns)


def _is_numeric_column(dataset: Dataset, name: str) -> bool:
    return any(c.name == name and c.type == "number" for c in dataset.columns)


def _validate_chart(chart: dict[str, Any], dataset: Dataset) -> bool:
    chart_type = chart.get("type", "")
    if chart_type in ("bar", "line", "scatter", "area", "combo"):
        return _valid_column(dataset, chart.get("x")) and _valid_column(dataset, chart.get("y"))
    if chart_type == "pie":
        return _valid_column(dataset, chart.get("category")) and _valid_column(dataset, chart.get("value"))
    if chart_type in ("histogram", "boxplot"):
        return _valid_column(dataset, chart.get("x")) and _is_numeric_column(dataset, chart.get("x", ""))
    if chart_type == "bubble":
        return (
            _valid_column(dataset, chart.get("x"))
            and _valid_column(dataset, chart.get("y"))
            and _valid_column(dataset, chart.get("size"))
        )
    if chart_type == "correlation":
        cols = chart.get("columns") or []
        return len(cols) >= 2 and all(_valid_column(dataset, c) and _is_numeric_column(dataset, c) for c in cols)
    return False


# ---------------------------------------------------------------------------
# Chart data generators
# ---------------------------------------------------------------------------

def _sample_rows(rows: list[dict], max_rows: int) -> list[dict]:
    if len(rows) <= max_rows:
        return rows
    step = len(rows) // max_rows
    return rows[::step][:max_rows]


def _generate_xy_data(dataset: Dataset, chart: dict) -> list[dict[str, Any]]:
    x_col, y_col = chart["x"], chart["y"]
    rows = _sample_rows(dataset.rows, _MAX_ROWS_FOR_CHART)
    data = []
    for row in rows:
        x_val = row.get(x_col)
        y_val = row.get(y_col)
        if x_val is None or y_val is None or x_val == "" or y_val == "":
            continue
        try:
            data.append({x_col: x_val, y_col: float(y_val)})
        except (ValueError, TypeError):
            data.append({x_col: x_val, y_col: y_val})
    return data


def _generate_pie_data(dataset: Dataset, chart: dict) -> list[dict[str, Any]]:
    cat_col, val_col = chart["category"], chart["value"]
    agg: dict[str, float] = defaultdict(float)
    for row in dataset.rows:
        cat = str(row.get(cat_col, ""))
        val = row.get(val_col)
        if cat and val is not None and val != "":
            try:
                agg[cat] += float(val)
            except (ValueError, TypeError):
                agg[cat] += 1
    # Top 10 categories
    sorted_items = sorted(agg.items(), key=lambda kv: kv[1], reverse=True)[:10]
    return [{cat_col: k, val_col: v} for k, v in sorted_items]


def _generate_histogram_data(dataset: Dataset, chart: dict) -> list[dict[str, Any]]:
    x_col = chart["x"]
    values = dataset.get_numeric_column(x_col)
    if not values:
        return []
    arr = np.array(values)
    n_bins = min(20, max(5, int(math.sqrt(len(arr)))))
    counts, bin_edges = np.histogram(arr, bins=n_bins)
    data = []
    for i, count in enumerate(counts):
        bin_start = round(float(bin_edges[i]), 4)
        bin_end = round(float(bin_edges[i + 1]), 4)
        data.append({"bin": f"{bin_start}-{bin_end}", "count": int(count), x_col: bin_start})
    return data


def _generate_boxplot_data(dataset: Dataset, chart: dict) -> list[dict[str, Any]]:
    x_col = chart["x"]
    values = dataset.get_numeric_column(x_col)
    if not values:
        return []
    arr = np.array(sorted(values))
    q1 = float(np.percentile(arr, 25))
    median = float(np.percentile(arr, 50))
    q3 = float(np.percentile(arr, 75))
    iqr = q3 - q1
    whisker_low = float(np.min(arr[arr >= q1 - 1.5 * iqr]))
    whisker_high = float(np.max(arr[arr <= q3 + 1.5 * iqr]))
    outliers = arr[(arr < whisker_low) | (arr > whisker_high)].tolist()
    return [{
        "column": x_col,
        "min": whisker_low,
        "q1": q1,
        "median": median,
        "q3": q3,
        "max": whisker_high,
        "outliers": outliers[:50],
    }]


def _generate_bubble_data(dataset: Dataset, chart: dict) -> list[dict[str, Any]]:
    x_col, y_col, size_col = chart["x"], chart["y"], chart["size"]
    rows = _sample_rows(dataset.rows, _MAX_ROWS_FOR_CHART)
    data = []
    for row in rows:
        try:
            x = float(row.get(x_col, ""))
            y = float(row.get(y_col, ""))
            sz = float(row.get(size_col, ""))
            data.append({x_col: x, y_col: y, size_col: sz})
        except (ValueError, TypeError):
            continue
    return data


def _generate_correlation_data(dataset: Dataset, chart: dict) -> list[dict[str, Any]]:
    cols = chart.get("columns", [])
    matrix = []
    for col_a in cols:
        a_vals = dataset.get_numeric_column(col_a)
        for col_b in cols:
            b_vals = dataset.get_numeric_column(col_b)
            if not a_vals or not b_vals:
                corr = 0.0
            else:
                min_len = min(len(a_vals), len(b_vals))
                a_arr = np.array(a_vals[:min_len])
                b_arr = np.array(b_vals[:min_len])
                if np.std(a_arr) == 0 or np.std(b_arr) == 0:
                    corr = 0.0
                else:
                    corr = float(np.corrcoef(a_arr, b_arr)[0, 1])
            matrix.append({"x": col_a, "y": col_b, "value": round(corr, 4)})
    return matrix


def _generate_chart_data(dataset: Dataset, chart: dict) -> list[dict[str, Any]]:
    t = chart.get("type", "")
    if t in ("bar", "line", "scatter", "area", "combo"):
        return _generate_xy_data(dataset, chart)
    if t == "pie":
        return _generate_pie_data(dataset, chart)
    if t == "histogram":
        return _generate_histogram_data(dataset, chart)
    if t == "boxplot":
        return _generate_boxplot_data(dataset, chart)
    if t == "bubble":
        return _generate_bubble_data(dataset, chart)
    if t == "correlation":
        return _generate_correlation_data(dataset, chart)
    return []


# ---------------------------------------------------------------------------
# Main orchestration
# ---------------------------------------------------------------------------

async def process_analysis(dataset: Dataset, prompt: str) -> AnalyzeResponse:
    ctx = _build_dataset_context(dataset)

    # Step 1: text insights
    insights = await gemini.analyze(ctx, prompt)

    # Step 2: chart generation with retries
    charts: list[ChartConfig] = []
    chart_status = "not_feasible"
    chart_message = ""
    retry_attempts = 0

    for attempt in range(_MAX_RETRIES):
        retry_attempts = attempt + 1
        try:
            raw_charts = await gemini.generate_charts(ctx, prompt, insights)
            valid_charts = []
            for rc in raw_charts:
                if not isinstance(rc, dict):
                    continue
                if not _validate_chart(rc, dataset):
                    logger.warning("Invalid chart config skipped: %s", rc)
                    continue
                rc["data"] = _generate_chart_data(dataset, rc)
                valid_charts.append(ChartConfig(**rc))

            if valid_charts:
                charts = valid_charts
                chart_status = "success" if len(valid_charts) == len(raw_charts) else "partial"
                chart_message = (
                    "Charts generated successfully."
                    if chart_status == "success"
                    else f"{len(valid_charts)} of {len(raw_charts)} charts were valid."
                )
                break
            else:
                chart_message = f"Attempt {retry_attempts}: no valid charts returned."
        except Exception as exc:
            chart_message = f"Attempt {retry_attempts} failed: {exc}"
            logger.warning("Chart generation attempt %d failed: %s", retry_attempts, exc)

    if not charts:
        chart_status = "failed"
        if not chart_message:
            chart_message = "Chart generation failed after all retries."

    return AnalyzeResponse(
        **{
            "insights": insights,
            "charts": charts,
            "chartStatus": chart_status,
            "chartMessage": chart_message,
            "retryAttempts": retry_attempts,
        }
    )
